### DISCUSSION TITLE
Detecting IDE mode

### DISCUSSION THREAD

**Discussion by BMOC1** [micropython/micropython#11626] [Using MicroPython]
Is there a way of programmatically knowing if im running code initiated through Thonny ie(IDE)...  Right now, i have a constant variable where i set inIDE to true or false.  If false, i do machine.reset() that i cannot do in IDE..

Being i have to rename the file to main.py to run automatically... I was thinking of just detecting the running script name.  But, i cannot find the micropython way of doing that.    Or maybe there is a better way.

In another words, if im in the IDE, i just re-initialize variables, but if I'm not in the IDE, i want to reboot/restart the device.

**Comment by dlech** [comment]
Depending on how the file is run (import vs. direct), you may be able to use `__name__ == ...` to tell the difference. Or if the board supports it, `__file__ == ...` could work too.

**Reply by BMOC1** [reply]
I had already tried both those variables - Under various attempts, i couldn't get` __file__` as a recognized declared variable?  the `__name__` returned **main** even though i was running a script not named main....

Im using a pico w.

**Comment by ned-pcs** [comment]
You can examine `machine.reset_cause()` to see.
If it is `machine.SOFT_RESET` you're likely in Thonny; if it's `machine.HARD_RESET` or `machine.PWRON_RESET` you're likely not in Thonny (or have done a `machine.reset()` from the REPL).

**Comment by aivarannamaa** [comment]
I have an idea.

Thonny represents running editor content with "magic" command `%Run -c $EDITOR_CONTENT` in the Shell.

The `-c` is inspired by similar CPython command line parameter. In CPython, when you do `python3 -c "print('Hello world!')"`, the first item in `sys.argv` becomes `"-c"`.

I could make Thonny to change `sys.argv` to `["-c"]` before it submits the editor content to the MicroPython REPL.
